src/sdf/src/sdf.cpp

```diff
@@ -356,5 +361,156 @@ Mesh LevelSet(std::function<float(glm::vec3)> sdf, Box bounds, float edgeLength,
+ * active.
+ * @return Mesh This class does not depend on Manifold, so it just returns a
+ * Mesh, but it is guaranteed to be manifold and so can always be used as
+ * input to the Manifold constructor for further operations.
+ */
+Mesh LevelSetBatch(
+    std::function<std::vector<float>(std::vector<glm::vec3>)> sdf, Box bounds,
+    float edgeLength, float level, bool canParallel) {
+  Mesh out;
+
+  const glm::vec3 dim = bounds.Size();
+  const float maxDim = std::max(dim[0], std::max(dim[1], dim[2]));
+  const glm::ivec3 gridSize(dim / edgeLength);
+  const glm::vec3 spacing = dim / (glm::vec3(gridSize));
+
+  const Uint64 maxMorton = MortonCode(glm::ivec4(gridSize + 1, 1));
+
+  // Parallel policies violate will crash language runtimes with runtime locks
+  // that expect to not be called back by unregistered threads. This allows
+  // bindings use LevelSet despite being compiled with MANIFOLD_PAR
+  // active.
+  const auto pol = canParallel ? autoPolicy(maxMorton) : ExecutionPolicy::Seq;
+
+  int tableSize = glm::min(
+      2 * maxMorton, static_cast<Uint64>(10 * glm::pow(maxMorton, 0.667)));
+  HashTable<GridVert, identity> gridVerts(tableSize);
+  Vec<glm::vec3> vertPos(gridVerts.Size() * 7, glm::vec3(0.0));
+
+  // BEGIN DIVERGENCE ----------------------------------------------------------
+
+  // Precompute the Coordinates for the desired Morton Indices
+  std::vector<glm::vec3> gridCoordinates(maxMorton + 1, glm::vec3(0.0));
+  for (Uint64 i = 0; i < maxMorton + 1; i++) {
+    const glm::ivec4 gridIndex = DecodeMorton(i);
+    // TODO: Handle grid indices outside of the traditional bounds...
+    // Indexing gets tough when coordinates are discarded...
+    // This halves the number of samples!!!
+    // assert(!(glm::any(glm::greaterThan(glm::ivec3(gridIndex), gridSize))));
+    gridCoordinates[i] = Position(gridIndex, bounds.min, spacing);
+  }
+
+  // Compute the distances at these coordinates
+  std::vector<float> gridDistances = sdf(gridCoordinates);
+
+  // Bound these distances - equivalent to BoundedSDF
+  for (Uint64 i = 0; i < maxMorton + 1; i++) {
+    const float d = gridDistances[i] - level;
+    const glm::ivec4 gridIndex = DecodeMorton(i);
+    const glm::ivec3 xyz(gridIndex);
+    const bool onLowerBound = glm::any(glm::lessThanEqual(xyz, glm::ivec3(0)));
+    const bool onUpperBound =
+        glm::any(glm::greaterThanEqual(xyz, (gridSize + 1)));
+    const bool onHalfBound = gridIndex.w == 1 && glm::any(glm::greaterThanEqual(
+                                                     xyz, (gridSize + 1) - 1));
+    gridDistances[i] =
+        (onLowerBound || onUpperBound || onHalfBound) ? glm::min(d, 0.0f) : d;
+  }
+
+  // Retry Implicit Construction with progressively larger and larger hashtables
+  while (1) {
+    // Check each grid coordinate to see if its neighbors cross the 0 threshold
+    size_t vertIndex = 0;
+    Uint64 lastMorton = 0;
+    for (Uint64 i = 0; i < maxMorton + 1; i++) {
```

Comment:
is there a reason this is sequential? considering the sdf invocation (the thing that cannot be called in parallel) is done once only, the remaining stuff should be parallelizable.

Reply:
Nope!  In theory, all the non-SDF call things can be done in parallel (but my intuition was that all the non-SDF stuff is negligible compared to the SDF evaluation... perhaps I'm mistaken!)

Reply:
It depends on the grid size I guess. If you do something like a million points, the non-SDF stuff is still pretty time consuming.

In our SDF examples, the SDF we use are pretty simple so the performance largely depends on the non-SDF stuff. The time difference between parallel build and sequential build in that case is still pretty large.